synapseclient/core/sts_transfer.py

```diff
@@ -0,0 +1,210 @@
+import collections
+import collections.abc
+import datetime
+import importlib
+import os
+import threading
+import platform
+
+from synapseclient.core.utils import iso_to_datetime, snake_case
+
+try:
+    boto3 = importlib.import_module('boto3')
+except ImportError:
+    # boto is not a requirement to load this module,
+    # we are able to optionally use functionality if it's available
+    boto3 = None
+
+STS_PERMISSIONS = set(['read_only', 'read_write'])
+
+
+class _TokenCache(collections.OrderedDict):
+    """A self pruning dictionary of STS tokens.
+    It will prune itself as new keys are added, removing the oldest if the
+    max_size is exceeded, and always removing all tokens that have expired
+    on each additional insert."""
+
+    def __init__(self, max_size):
+        super().__init__()
+        self.max_size = max_size
+
+    def __setitem__(self, key, value):
+        super().__setitem__(key, value)
+        self._prune()
+
```

Comment:
It's not a FIFO cache since it is implicitly ordered by the expiration of the tokens. If it were an LRU cache then it could expire a token with longer life that hadn't been more recently used before a token that was about to expire. It's not necessarily obvious to me that that is preferable.

Reply:
It will function in the edge case just generating new tokens and re-adding them, as you say. It will never not return a token for the requested location due to being full. If it were to switch to FIFO then you'd need to inspect all the tokens every prune to see if they had expired, not just check until you found one that hadn't (or additionally keep a separate data structure).